server/app/services/question/ActiveAndDraftQuestions.java

```diff
@@ -24,39 +33,95 @@ public final class ActiveAndDraftQuestions {
           String, Pair<Optional<QuestionDefinition>, Optional<QuestionDefinition>>>
       versionedByName;
   private final ImmutableMap<String, DeletionStatus> deletionStatusByName;
+  private final Optional<ImmutableMap<String, ImmutableSet<ProgramDefinition>>>
+      referencingDraftProgramsByName;
+  private final ImmutableMap<String, ImmutableSet<ProgramDefinition>>
+      referencingActiveProgramsByName;
 
-  public ActiveAndDraftQuestions(Version active, Version draft) {
-    ImmutableMap.Builder<String, QuestionDefinition> activeToName = ImmutableMap.builder();
-    ImmutableMap.Builder<String, QuestionDefinition> draftToName = ImmutableMap.builder();
-    ImmutableMap.Builder<String, DeletionStatus> deletionStatusBuilder = ImmutableMap.builder();
-    draft.getQuestions().stream()
-        .map(Question::getQuestionDefinition)
-        .forEach(qd -> draftToName.put(qd.getName(), qd));
-    active.getQuestions().stream()
-        .map(Question::getQuestionDefinition)
-        .forEach(qd -> activeToName.put(qd.getName(), qd));
-    ImmutableMap<String, QuestionDefinition> activeNames = activeToName.build();
-    ImmutableMap<String, QuestionDefinition> draftNames = draftToName.build();
-    ImmutableMap.Builder<String, Pair<Optional<QuestionDefinition>, Optional<QuestionDefinition>>>
-        versionedByNameBuilder = ImmutableMap.builder();
-    for (String name : Sets.union(activeNames.keySet(), draftNames.keySet())) {
-      versionedByNameBuilder.put(
-          name,
-          Pair.create(
-              Optional.ofNullable(activeNames.get(name)),
-              Optional.ofNullable(draftNames.get(name))));
+  public ActiveAndDraftQuestions(VersionRepository repository) {
+    Version active = repository.getActiveVersion();
+    Version draft = repository.getDraftVersion();
```

Comment:
suggestion (blocking): c'tors generally should just initialize resources, but shouldn't themselves do much logic or any i/o. The previous version of this code wasn't great, but rather than moving it toward doing performing queries in the c'tor I recommend moving the query and collation logic into a private method that gets called in the public methods if it hasn't run yet i.e. the optional values it sets are empty.

Reply:
Doing the IO in the static method is definitely an improvement, and way less of a lift in comparison to rewriting the whole thing to be lazy.

**I'm ok with merging as-is.**

That being said, to address the multithreading issue you brought up, I think that could be resolved by making the private method that performs the I/O and collates the results [`synchronized`](https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html).

Something like:

```java
  private synchronized void ensureResults() {
    // load data from DB, collate it and set the variables
  }
  
  public boolean draftVersionHasAnyEdits() {
    ensureResults();
    return draftVersionHasAnyEdits;
  }
```

This wouldn't address your concern of reduced immutability though, though the mutability would be internal only.

Reply:
Sounds good. I'm going to merge this as-is and try and put together a follow-on PR that moves all initialization of this data to an actual function that initializes a plain-old-data-object.